File: src/project/gitignore.ts
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

/**
 * Minimal .gitignore matcher.
 *
 * Supports the common subset of gitignore syntax:
 *  - blank lines and `#` comments
 *  - `!` negation
 *  - trailing `/` (directory-only)
 *  - leading `/` (anchored to the .gitignore location)
 *  - patterns containing `/` are anchored to the .gitignore location
 *  - patterns without `/` match the basename at any depth
 *  - `*`, `?`, `**`, `[...]` globs
 */
export interface IgnoreRule {
  /** Compiled matcher. */
  regex: RegExp;
  /** True when this rule re-includes files (`!pattern`). */
  negated: boolean;
  /** True when this rule only matches directories. */
  dirOnly: boolean;
  /** True when the pattern is anchored to its .gitignore location. */
  anchored: boolean;
  /** Relative dir (from project root, '' for root) containing the .gitignore. */
  baseRel: string;
}

const ALWAYS_IGNORED = new Set(['.git', 'node_modules', '.zeesh', '.myagent']); // .myagent = pre-rename state dir

function globToRegExp(pattern: string): RegExp {
  let out = '';
  let i = 0;
  while (i < pattern.length) {
    const ch = pattern[i] as string;
    if (ch === '*') {
      if (pattern[i + 1] === '*') {
        // `**` — matches anything including slashes
        if (pattern[i + 2] === '/') {
          out += '(?:.*/)?';
          i += 3;
        } else {
          out += '.*';
          i += 2;
        }
      } else {
        out += '[^/]*';
        i += 1;
      }
    } else if (ch === '?') {
      out += '[^/]';
      i += 1;
    } else if (ch === '[') {
      // character class — copy until closing bracket
      let j = i + 1;
      let cls = '';
      if (pattern[j] === '!' || pattern[j] === '^') {
        cls += '^';
        j += 1;
      }
      while (j < pattern.length && pattern[j] !== ']') {
        if (pattern[j] === '\\') {
          cls += '\\\\' + (pattern[j + 1] ?? '');
          j += 2;
        } else {
          cls += pattern[j];
          j += 1;
        }
      }
      if (j >= pattern.length) {
        out += '\\[';
        i += 1;
      } else {
        out += '[' + cls + ']';
        i = j + 1;
      }
    } else {
      out += ch.replace(/[.+^${}()|\\]/g, '\\$&');
      i += 1;
    }
  }
  return new RegExp('^' + out + '$');
}

/** Parse the contents of a .gitignore located at `baseRel` ('' for project root). */
export function parseGitignore(content: string, baseRel: string): IgnoreRule[] {
  const rules: IgnoreRule[] = [];
  for (const rawLine of content.split(/\r?\n/)) {
    let line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;
    if (line.startsWith('\\#')) line = line.slice(1);

    let negated = false;
    if (line.startsWith('!')) {
      negated = true;
      line = line.slice(1);
    }
    if (!line) continue;

    // Trailing spaces are stripped unless escaped
    line = line.replace(/\s+$/, '');

    let dirOnly = false;
    if (line.endsWith('/')) {
      dirOnly = true;
      line = line.slice(0, -1);
    }
    if (!line) continue;

    let anchored = false;
    if (line.startsWith('/')) {
      anchored = true;
      line = line.slice(1);
    } else if (line.includes('/')) {
      anchored = true;
    }

    try {
      rules.push({ regex: globToRegExp(line), negated, dirOnly, anchored, baseRel });
    } catch {
      // Skip malformed patterns rather than crashing the walker
    }
  }
  return rules;
}

/** Load .gitignore rules from a directory (returns [] when absent). */
export function loadGitignoreRules(dir: string, baseRel: string): IgnoreRule[] {
  try {
    return parseGitignore(readFileSync(join(dir, '.gitignore'), 'utf8'), baseRel);
  } catch {
    return [];
  }
}

/** Normalize a relative path to forward slashes without a leading slash. */
export function normalizeRel(p: string): string {
  return p.replace(/\\/g, '/').replace(/^\.?\//, '').replace(/^\/+/, '');
}

/**
 * Decide whether `rel` (relative path from project root, forward slashes)
 * is ignored by the given rules. Returns true when ignored.
 *
 * Directory-only rules also match files under the matched directory, so a
 * rule like `dist/` ignores `dist/bundle.js` as well as `dist/` itself.
 */
export function isIgnored(rel: string, isDir: boolean, rules: IgnoreRule[]): boolean {
  const normalized = normalizeRel(rel);
  const segments = normalized.split('/');

  if (segments.some((s) => ALWAYS_IGNORED.has(s))) return true;

  // Ancestor directory prefixes (for directory-only rules). Includes the
  // path itself when it is a directory.
  const dirPrefixes: string[] = [];
  for (let i = 1; i < segments.length; i++) {
    dirPrefixes.push(segments.slice(0, i).join('/'));
  }
  if (isDir) dirPrefixes.push(normalized);

  let ignored = false;
  for (const rule of rules) {
    if (rule.dirOnly) {
      // Matches when any ancestor directory (or the dir itself) matches.
      for (const dir of dirPrefixes) {
        const relToBase = applyBase(rule, dir);
        if (relToBase === '') continue;
        const target = rule.anchored ? relToBase : lastSegment(relToBase);
        if (rule.regex.test(target)) {
          ignored = !rule.negated;
          break;
        }
      }
    } else {
      const relToBase = applyBase(rule, normalized);
      if (relToBase === '') continue;
      const target = rule.anchored ? relToBase : lastSegment(relToBase);
      if (rule.regex.test(target)) ignored = !rule.negated;
    }
  }
  return ignored;
}

/** Reduce `path` relative to a rule's .gitignore location ('' when out of scope). */
function applyBase(rule: IgnoreRule, path: string): string {
  if (rule.baseRel === '') return path;
  const prefix = normalizeRel(rule.baseRel) + '/';
  return path.startsWith(prefix) ? path.slice(prefix.length) : '';
}

function lastSegment(path: string): string {
  const segs = path.split('/');
  return segs[segs.length - 1] as string;
}

File: grace/project/gitignore.py
"""Minimal .gitignore matcher (port of src/project/gitignore.ts).

Supports the common subset of gitignore syntax:
 - blank lines and `#` comments
 - `!` negation
 - trailing `/` (directory-only)
 - leading `/` (anchored to the .gitignore location)
 - patterns containing `/` are anchored to the .gitignore location
 - patterns without `/` match the basename at any depth
 - `*`, `?`, `**`, `[...]` globs
"""

import os
import re
from dataclasses import dataclass

ALWAYS_IGNORED = {".git", "node_modules", ".zeesh", ".myagent"}


@dataclass
class IgnoreRule:
    regex: re.Pattern
    negated: bool
    dirOnly: bool
    anchored: bool
    baseRel: str


def _glob_to_regexp(pattern: str) -> re.Pattern:
    out = ""
    i = 0
    while i < len(pattern):
        ch = pattern[i]
        if ch == "*":
            if i + 1 < len(pattern) and pattern[i + 1] == "*":
                if i + 2 < len(pattern) and pattern[i + 2] == "/":
                    out += "(?:.*/)?"
                    i += 3
                else:
                    out += ".*"
                    i += 2
            else:
                out += "[^/]*"
                i += 1
        elif ch == "?":
            out += "[^/]"
            i += 1
        elif ch == "[":
            # character class — copy until closing bracket
            j = i + 1
            cls = ""
            if j < len(pattern) and pattern[j] in ("!", "^"):
                cls += "^"
                j += 1
            while j < len(pattern) and pattern[j] != "]":
                if pattern[j] == "\\":
                    cls += "\\\\" + (pattern[j + 1] if j + 1 < len(pattern) else "")
                    j += 2
                else:
                    cls += pattern[j]
                    j += 1
            if j >= len(pattern):
                out += "\\["
                i += 1
            else:
                out += "[" + cls + "]"
                i = j + 1
        else:
            out += re.escape(ch)
            i += 1
    return re.compile("^" + out + "$")


def parse_gitignore(content: str, base_rel: str) -> list[IgnoreRule]:
    rules: list[IgnoreRule] = []
    for raw_line in content.splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        if line.startswith("\\#"):
            line = line[1:]

        negated = False
        if line.startswith("!"):
            negated = True
            line = line[1:]
        if not line:
            continue

        # Trailing spaces are stripped unless escaped
        line = re.sub(r"\s+$", "", line)

        dir_only = False
        if line.endswith("/"):
            dir_only = True
            line = line[:-1]
        if not line:
            continue

        anchored = False
        if line.startswith("/"):
            anchored = True
            line = line[1:]
        elif "/" in line:
            anchored = True

        try:
            rules.append(IgnoreRule(regex=_glob_to_regexp(line), negated=negated, dirOnly=dir_only, anchored=anchored, baseRel=base_rel))
        except re.error:
            # Skip malformed patterns rather than crashing the walker
            pass
    return rules


def load_gitignore_rules(dir_path: str, base_rel: str) -> list[IgnoreRule]:
    try:
        with open(os.path.join(dir_path, ".gitignore"), encoding="utf-8") as f:
            return parse_gitignore(f.read(), base_rel)
    except OSError:
        return []


def normalize_rel(p: str) -> str:
    p = p.replace("\\", "/")
    p = re.sub(r"^\.?/", "", p)
    p = re.sub(r"^/+", "", p)
    return p


def is_ignored(rel: str, is_dir: bool, rules: list[IgnoreRule]) -> bool:
    normalized = normalize_rel(rel)
    segments = normalized.split("/")

    if any(s in ALWAYS_IGNORED for s in segments):
        return True

    # Ancestor directory prefixes (for directory-only rules). Includes the
    # path itself when it is a directory.
    dir_prefixes = []
    for i in range(1, len(segments)):
        dir_prefixes.append("/".join(segments[:i]))
    if is_dir:
        dir_prefixes.append(normalized)

    ignored = False
    for rule in rules:
        if rule.dirOnly:
            for dir_prefix in dir_prefixes:
                rel_to_base = _apply_base(rule, dir_prefix)
                if rel_to_base == "":
                    continue
                target = rel_to_base if rule.anchored else _last_segment(rel_to_base)
                if rule.regex.search(target):
                    ignored = not rule.negated
                    break
        else:
            rel_to_base = _apply_base(rule, normalized)
            if rel_to_base == "":
                continue
            target = rel_to_base if rule.anchored else _last_segment(rel_to_base)
            if rule.regex.search(target):
                ignored = not rule.negated
    return ignored


def _apply_base(rule: IgnoreRule, path: str) -> str:
    if rule.baseRel == "":
        return path
    prefix = normalize_rel(rule.baseRel) + "/"
    return path[len(prefix):] if path.startswith(prefix) else ""


def _last_segment(path: str) -> str:
    segs = path.split("/")
    return segs[-1] if segs else ""
